Port AXID to ObjectIdentifier

cdumez opened the pull request:
#### 42ed41c5d33eca7d0b52ed9aead4c2cdde2880c1
<pre>
Port AXID to ObjectIdentifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=281390">https://bugs.webkit.org/show_bug.cgi?id=281390</a>

Reviewed by Darin Adler.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isTableCellInSameRowGroup):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::objectID const):
(WebCore::AXCoreObject::rowGroupAncestorID const):
(WebCore::AXCoreObject::AXCoreObject):
(WebCore::axIDs):
(WebCore::Accessibility::retrieveValueFromMainThread):
* Source/WebCore/accessibility/AXGeometryManager.cpp:
(WebCore::AXGeometryManager::cacheRect):
* Source/WebCore/accessibility/AXGeometryManager.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
(WebCore::streamAXCoreObject):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::setIsolatedTreeFocusedObject):
(WebCore::AXObjectCache::get const):
(WebCore::AXObjectCache::cacheAndInitializeWrapper):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::generateNewObjectID const):
(WebCore::AXObjectCache::getAXID):
(WebCore::AXObjectCache::onRendererCreated):
(WebCore::AXObjectCache::visiblePositionForTextMarkerData):
(WebCore::AXObjectCache::objectForTextMarkerData):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::updateIsolatedTree const):
(WebCore::AXObjectCache::onPaint const):
(WebCore::AXObjectCache::addRelation):
(WebCore::AXObjectCache::removeRelation):
(WebCore::AXObjectCache::isDescendantOfRelatedNode):
(WebCore::AXObjectCache::relatedObjectIDsFor):
(WebCore::AXObjectCache::selectedTextRangeTimerFired):
* Source/WebCore/accessibility/AXObjectCache.h:
(WebCore::AXObjectCache::nodeForID const):
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::matchForSearchKeyAtIndex):
(WebCore::AXSearchManager::findMatchingRange):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::nodeID):
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::AXTextMarker::operator CharacterOffset const):
(WebCore::AXTextMarker::object const):
(WebCore::AXTextMarker::debugDescription const):
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::AXTextMarkerRange::isConfinedTo const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::TextMarkerData::axTreeID const):
(WebCore::TextMarkerData::axObjectID const):
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::treeID const):
(WebCore::AXTextMarker::objectID const):
(WebCore::AXTextMarker::isNull const):
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
* Source/WebCore/accessibility/AXTreeStore.h:
(WebCore::AXTreeStore&lt;T&gt;::axObjectCacheForID):
(WebCore::AXTreeStore&lt;T&gt;::isolatedTreeForID):
(WebCore::AXTreeStore&lt;T&gt;::generateNewID):
(WebCore::axTreeForID):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::treeID const):
(WebCore::AccessibilityObject::dbg const):
(WebCore::AccessibilityObject::isIgnored const):
(WebCore::AccessibilityObject::isIgnoredWithoutCache const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::cellSlots):
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::cellSlots):
(WebCore::AccessibilityTable::addChildren):
(WebCore::AccessibilityTable::cellForColumnAndRow):
* Source/WebCore/accessibility/AccessibilityTable.h:
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::rowGroupAncestorID const):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::AXIsolatedObject):
(WebCore::AXIsolatedObject::dbg const):
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::associatedAXObject const):
(WebCore::AXIsolatedObject::setMathscripts):
(WebCore::AXIsolatedObject::setObjectProperty):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::isDetachedFromParent):
(WebCore::AXIsolatedObject::cellForColumnAndRow):
(WebCore::AXIsolatedObject::accessibilityHitTest const):
(WebCore::AXIsolatedObject::objectAttributeValue const):
(WebCore::AXIsolatedObject::textInputMarkedTextMarkerRange const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::createEmptyContent):
(WebCore::AXIsolatedTree::reportLoadingProgress):
(WebCore::AXIsolatedTree::objectForID const):
(WebCore::AXIsolatedTree::shouldCreateNodeChange):
(WebCore::AXIsolatedTree::nodeChangeForObject):
(WebCore::AXIsolatedTree::queueChange):
(WebCore::AXIsolatedTree::addUnconnectedNode):
(WebCore::AXIsolatedTree::queueAppendsAndRemovals):
(WebCore::AXIsolatedTree::collectNodeChangesForSubtree):
(WebCore::AXIsolatedTree::updateNode):
(WebCore::AXIsolatedTree::updatePropertiesForSelfAndDescendants):
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::AXIsolatedTree::updateDependentProperties):
(WebCore::AXIsolatedTree::updateChildren):
(WebCore::AXIsolatedTree::focusedNodeID):
(WebCore::AXIsolatedTree::setFocusedNodeID):
(WebCore::AXIsolatedTree::removeNode):
(WebCore::AXIsolatedTree::removeSubtreeFromNodeMap):
(WebCore::AXIsolatedTree::relatedObjectIDsFor):
(WebCore::AXIsolatedTree::applyPendingChanges):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(-[WebAccessibilityObjectWrapperBase attachAXObject:]):
(-[WebAccessibilityObjectWrapperBase attachIsolatedObject:]):
(-[WebAccessibilityObjectWrapperBase detach]):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/285102@main">https://commits.webkit.org/285102@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3abc29771a54e54b2d75da7f3fedbc1c2666e6ac

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71546 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50959 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24320 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75658 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22751 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58766 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22571 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56506 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14978 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74612 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61625 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36955 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42899 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19091 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21092 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64802 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19455 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77376 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15780 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18630 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64221 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15823 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61664 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64216 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12361 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5999 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10967 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46759 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1538 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47830 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49114 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47572 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -365,8 +365,8 @@ bool AXCoreObject::isTableCellInSameRowGroup(AXCoreObject* otherTableCell)
-    AXID ancestorID = rowGroupAncestorID();
-    return ancestorID.isValid() && ancestorID == otherTableCell->rowGroupAncestorID();
+    auto ancestorID = rowGroupAncestorID();
+    return ancestorID && *ancestorID == otherTableCell->rowGroupAncestorID();
```
I think this can be
```
return ancestorID == otherTableCell->rowGroupAncestorID();
```
That's how `std::optional` `==` works.

cdumez replied:
Not quite though. `rowGroupAncestorID()` returns an optional. The pre-existing code used to return `false` if both `rowGroupAncestorID()` and `otherTableCell->rowGroupAncestorID()` were `std::nullopt`. With your proposal it would start returning true.

I am not familiar with this code so I am not sure it matters but I opted to match the pre-existing behavior as closely as possible to limit risk.

darinadler replied:
Got it.

darinadler commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -1412,8 +1412,8 @@ class AXCoreObject : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<AXCo
 {
-    return objects.map([] (const auto& object) {
-        return object ? object->objectID() : AXID();
+    return WTF::compactMap(objects, [](const auto& object) {
```
This is a behavior change. The old version would have nulls in the vector, I think, and the new one omits them. Is that OK?

cdumez replied:
Call sites expects the IDs to be valid. Maybe I can use map() instead of compactMap() and assert that the IDs are valid for clarity.

darinadler commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1151,33 +1143,33 @@ void AXObjectCache::remove(Widget* view)
+        return *objectID;
     }
 
     objectID = generateNewObjectID();
```
Could use a separate local for this, which would be non-optional.

darinadler commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1151,33 +1143,33 @@ void AXObjectCache::remove(Widget* view)
-    if (objectID.isValid()) {
-        ASSERT(m_idsInUse.contains(objectID));
-        return objectID;
+    auto objectID = object.objectID();
```
Could put this inside the `if` so it's scoped.

darinadler commented on Source/WebCore/accessibility/AXTextMarker.h:
```diff
@@ -162,10 +162,10 @@ class AXTextMarker {
+    std::optional<AXID> objectID() const { return m_data.axObjectID(); }
     unsigned offset() const { return m_data.offset; }
-    bool isNull() const { return !treeID().isValid() || !objectID().isValid(); }
+    bool isNull() const { return !treeID() || !objectID(); }
```
This sure is a strange definition of null, but not new.

darinadler commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -817,7 +818,7 @@ void AXIsolatedTree::updateChildren(AccessibilityObject& axObject, ResolveNodeCh
     // an associated isolated object and update its children.
     auto* axAncestor = Accessibility::findAncestor(axObject, true, [this] (auto& ancestor) {
-        return m_nodeMap.find(ancestor.objectID()) != m_nodeMap.end();
+        return m_nodeMap.find(*ancestor.objectID()) != m_nodeMap.end();
```
Surprised there is not a `contains` function to use here.

cdumez committed:
Port AXID to ObjectIdentifier
https://bugs.webkit.org/show_bug.cgi?id=281390

Reviewed by Darin Adler.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isTableCellInSameRowGroup):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::objectID const):
(WebCore::AXCoreObject::rowGroupAncestorID const):
(WebCore::AXCoreObject::AXCoreObject):
(WebCore::axIDs):
(WebCore::Accessibility::retrieveValueFromMainThread):
* Source/WebCore/accessibility/AXGeometryManager.cpp:
(WebCore::AXGeometryManager::cacheRect):
* Source/WebCore/accessibility/AXGeometryManager.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
(WebCore::streamAXCoreObject):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::setIsolatedTreeFocusedObject):
(WebCore::AXObjectCache::get const):
(WebCore::AXObjectCache::cacheAndInitializeWrapper):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::generateNewObjectID const):
(WebCore::AXObjectCache::getAXID):
(WebCore::AXObjectCache::onRendererCreated):
(WebCore::AXObjectCache::visiblePositionForTextMarkerData):
(WebCore::AXObjectCache::objectForTextMarkerData):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::updateIsolatedTree const):
(WebCore::AXObjectCache::onPaint const):
(WebCore::AXObjectCache::addRelation):
(WebCore::AXObjectCache::removeRelation):
(WebCore::AXObjectCache::isDescendantOfRelatedNode):
(WebCore::AXObjectCache::relatedObjectIDsFor):
(WebCore::AXObjectCache::selectedTextRangeTimerFired):
* Source/WebCore/accessibility/AXObjectCache.h:
(WebCore::AXObjectCache::nodeForID const):
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::matchForSearchKeyAtIndex):
(WebCore::AXSearchManager::findMatchingRange):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::nodeID):
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::AXTextMarker::operator CharacterOffset const):
(WebCore::AXTextMarker::object const):
(WebCore::AXTextMarker::debugDescription const):
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::AXTextMarkerRange::isConfinedTo const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::TextMarkerData::axTreeID const):
(WebCore::TextMarkerData::axObjectID const):
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::treeID const):
(WebCore::AXTextMarker::objectID const):
(WebCore::AXTextMarker::isNull const):
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
* Source/WebCore/accessibility/AXTreeStore.h:
(WebCore::AXTreeStore<T>::axObjectCacheForID):
(WebCore::AXTreeStore<T>::isolatedTreeForID):
(WebCore::AXTreeStore<T>::generateNewID):
(WebCore::axTreeForID):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::treeID const):
(WebCore::AccessibilityObject::dbg const):
(WebCore::AccessibilityObject::isIgnored const):
(WebCore::AccessibilityObject::isIgnoredWithoutCache const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::cellSlots):
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::cellSlots):
(WebCore::AccessibilityTable::addChildren):
(WebCore::AccessibilityTable::cellForColumnAndRow):
* Source/WebCore/accessibility/AccessibilityTable.h:
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::rowGroupAncestorID const):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::AXIsolatedObject):
(WebCore::AXIsolatedObject::dbg const):
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::associatedAXObject const):
(WebCore::AXIsolatedObject::setMathscripts):
(WebCore::AXIsolatedObject::setObjectProperty):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::isDetachedFromParent):
(WebCore::AXIsolatedObject::cellForColumnAndRow):
(WebCore::AXIsolatedObject::accessibilityHitTest const):
(WebCore::AXIsolatedObject::objectAttributeValue const):
(WebCore::AXIsolatedObject::textInputMarkedTextMarkerRange const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::createEmptyContent):
(WebCore::AXIsolatedTree::reportLoadingProgress):
(WebCore::AXIsolatedTree::objectForID const):
(WebCore::AXIsolatedTree::shouldCreateNodeChange):
(WebCore::AXIsolatedTree::nodeChangeForObject):
(WebCore::AXIsolatedTree::queueChange):
(WebCore::AXIsolatedTree::addUnconnectedNode):
(WebCore::AXIsolatedTree::queueAppendsAndRemovals):
(WebCore::AXIsolatedTree::collectNodeChangesForSubtree):
(WebCore::AXIsolatedTree::updateNode):
(WebCore::AXIsolatedTree::updatePropertiesForSelfAndDescendants):
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::AXIsolatedTree::updateDependentProperties):
(WebCore::AXIsolatedTree::updateChildren):
(WebCore::AXIsolatedTree::focusedNodeID):
(WebCore::AXIsolatedTree::setFocusedNodeID):
(WebCore::AXIsolatedTree::removeNode):
(WebCore::AXIsolatedTree::removeSubtreeFromNodeMap):
(WebCore::AXIsolatedTree::relatedObjectIDsFor):
(WebCore::AXIsolatedTree::applyPendingChanges):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(-[WebAccessibilityObjectWrapperBase attachAXObject:]):
(-[WebAccessibilityObjectWrapperBase attachIsolatedObject:]):
(-[WebAccessibilityObjectWrapperBase detach]):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/285102@main
